File: translated/f-to-c/DRB016-outputdep-orig-yes.c
/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.

SPDX-License-Identifier: (BSD-3-Clause)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//The loop in this example cannot be parallelized.
//
//This pattern has two pair of dependencies:
//1. loop carried output dependence
// x = .. :
//
//2. loop carried true dependence due to:
//.. = x;
// x = ..;
//Data race pairs: we allow two pairs to preserve the original code pattern.
// 1. x@48:16:R vs. x@49:9:W
// 2. x@49:9:W vs. x@49:9:W

#include <omp.h>
#include <stdio.h>
#include <stdlib.h>

// Global array (from module)
int* a;

void useGlobalArray(int len)
{
    len = 100;
    a = (int*)malloc(100 * sizeof(int));
}

int main()
{
    int len, i, x;

    len = 100;
    x = 10;

    useGlobalArray(len);

    #pragma omp parallel for
    for (i = 0; i < len; i++) {
        a[i] = x;
        x = i + 1;  // Adjust for 0-based indexing in C
    }

    printf("x = %d\n", x);

    free(a);

    return 0;
} 

File: translated/f-to-go-automodel-bad/DRB016-outputdep-orig-yes.go
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Copyright (c) 2017-20, Lawrence Livermore National Security, LLC
// and DataRaceBench project contributors. See the DataRaceBench/COPYRIGHT file for details.
//
// SPDX-License-Identifier: (BSD-3-Clause)
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

//The loop in this example cannot be parallelized.
//
//This pattern has two pair of dependencies:
//1. loop carried output dependence
// x = .. :
//
//2. loop carried true dependence due to:
//.. = x;
// x = ..;
//Data race pairs: we allow two pairs to preserve the original code pattern.
// 1. x@48:16:R vs. x@49:9:W
// 2. x@49:9:W vs. x@49:9:W

package main

import (
	"fmt"
	"runtime"
	"sync"
)

// globalArray module equivalent
var a []int

func useGlobalArray(length *int) {
	*length = 100
	a = make([]int, 100)
}

func main() {
	var length, x int

	length = 100
	x = 10

	useGlobalArray(&length)

	//$omp parallel do
	var wg sync.WaitGroup
	numCPU := runtime.NumCPU()
	chunkSize := length / numCPU
	if chunkSize < 1 {
		chunkSize = 1
	}

	for start := 1; start <= length; start += chunkSize {
		end := start + chunkSize - 1
		if end > length {
			end = length
		}
		wg.Add(1)
		go func(start, end int) {
			defer wg.Done()
			for i := start; i <= end; i++ {
				a[i-1] = x
				x = i
			}
		}(start, end)
	}
	wg.Wait()
	//$omp end parallel do

	fmt.Printf("x = %d\n", x)
}
